src/ca/verify/signer.ts

```diff
@@ -0,0 +1,136 @@
+*/
+import { VerificationError } from '../../error';
+import * as sigstore from '../../types/sigstore';
+import { x509Certificate } from '../../x509/cert';
+
+// https://github.com/sigstore/fulcio/blob/main/docs/oid-info.md#1361415726411--issuer
+const OID_FUCIO_ISSUER = '1.3.6.1.4.1.57264.1.1';
+
+// Verifies the identity embedded in a Fulcio-issued signing certificate against
+// the list of trusted identities. Returns without error if at least one of the
+// identities matches the signing certificate; otherwise, throws a
+// VerificationError.
+export function verifySignerIdentity(
+  signingCert: x509Certificate,
+  identities: sigstore.CertificateIdentities
+): void {
+  // Check that the signing certificate was issued to at least one of the
+  // specified identities
+  const signerVerified = identities.identities.some((identity) =>
+    verifyIdentity(signingCert, identity)
+  );
+
+  if (!signerVerified) {
+    throw new VerificationError('Certificate issued to untrusted signer');
+  }
+}
+
+// Checks that the specified certificate was issued to the specified identity.
+// The certificate must match the issuer, subject alternative name, and an
+// optional list of certificate extensions. Returns true if the certificate was
+// issued to the identity; otherwise, returns false.
+function verifyIdentity(
+  cert: x509Certificate,
+  identity: sigstore.CertificateIdentity
+): boolean {
+  return (
+    verifyIssuer(cert, identity.issuer) &&
+    verifySAN(cert, identity.san) &&
+    verifyOIDs(cert, identity.oids)
+  );
+}
+
+// Checks the Fulcio issuer extension against the expected issuer. Returns true
+// if the issuer matches; otherwise, returns false.
+function verifyIssuer(cert: x509Certificate, issuer: string): boolean {
+  const issuerExtension = cert.extension(OID_FUCIO_ISSUER);
+
+  return issuerExtension?.value.toString('ascii') === issuer;
+}
+
+// Checks the certificate against the expected subject alternative name. Returns
+// true if the SAN matches; otherwise, returns false.
+function verifySAN(
+  cert: x509Certificate,
+  expectedSAN?: sigstore.SubjectAlternativeName
+): boolean {
+  // Fail if the SAN is not specified or is not a supported type
+  if (
+    expectedSAN === undefined ||
+    expectedSAN.identity === undefined ||
+    expectedSAN.type ===
+      sigstore.SubjectAlternativeNameType
+        .SUBJECT_ALTERNATIVE_NAME_TYPE_UNSPECIFIED
+  ) {
+    return false;
+  }
+
+  const sanExtension = cert.extSubjectAltName;
+
+  // Fail if the certificate does not have a SAN extension
+  if (!sanExtension) {
+    return false;
+  }
+
+  let sanValue: string | undefined;
+  switch (expectedSAN.type) {
+    case sigstore.SubjectAlternativeNameType.EMAIL:
+      sanValue = sanExtension.rfc822Name;
+      break;
+    case sigstore.SubjectAlternativeNameType.URI:
+      sanValue = sanExtension.uri;
+      break;
+    case sigstore.SubjectAlternativeNameType.OTHER_NAME:
+    // TODO: Support other name types
```

Comment:
Support for this is going to require some work in the `x509SubjectAlternativeNameExtension` class in order to retrieve the `OtherName` value (see [rfc5280](https://www.rfc-editor.org/rfc/rfc5280#section-4.2.1.6)). Gonna tackle this in a separate PR.